app/models/interactive_run_state.rb

```diff
@@ -114,22 +122,46 @@ def has_linked_interactive
     (interactive.respond_to? :linked_interactive) && !interactive.linked_interactive.nil?
   end
 
-  def linked_state
-    # Note that this method will return the first available state of the linked interactives *chain*.
-    # Sometimes the interactive that is directly linked might have not be run by student (e.g. when he skipped a page).
+  # Returns all linked states. 1st element in the array is an interactive directly linked to given one.
+  def all_linked_states(host = nil)
+    result = []
     current_interactive = interactive
     while (current_interactive.respond_to? :linked_interactive) && !current_interactive.linked_interactive.nil?
       linked_interactive = current_interactive.linked_interactive
       runs = run.sequence_run ? run.sequence_run.runs : [run]
       state = InteractiveRunState.where(run_id: runs.map(&:id), interactive_id: linked_interactive.id).first
-      if state && state.raw_data
-        return state.raw_data
-      else
-        # Try next interactive in chain.
-        current_interactive = linked_interactive
+      # It's not necessary to provide some metadata even for empty linked states, although it doesn't cost much
+      # and it can be useful for interactive itself to know that there are some linked interactives,
+      # but not run by student yet.
+      linked_state_info = { interactive_id: current_interactive.id, data: nil, created_at: nil, updated_at: nil }
```

Comment:
Shouldn't the id here be the `linked_interactive.id`?

Reply:
It should, thanks! Fixed.